### ISSUE TITLE
repository_ctx.download needs a fail-tolerant boolean

### ISSUE THREAD

**Issue by cgruber** [bazelbuild/bazel#6734]
### Description of the problem / feature request:

`repository_ctx.download()` currently fails when a download attempt runs through all the supplied urls.  This doesn't allow for more supple use-cases, such as downloading maven sources, which may or may not succeed, but is not fatal, since not all maven artifacts are deployed with sources.  

### Feature requests: what underlying problem are you trying to solve with this feature?

Having an option, such as a simple boolean "fatal" or some alternative signal to say what bazel's behavior should be, in a less critical case. 

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

create a simple rule to download a file that doesn't exist, execute it, and watch it fail. 

```
_BUILD_FILE_TEMPLATE = """
package(default_visibility = ["//visibility:public"])
exports(["{path}"])
"""

_DOWNLOAD_PREFIX = "maven"
def _fetch_file(ctx):
    repository_root_path = ctx.path(".")
    local_path = "%s/%s" % (_DOWNLOAD_PREFIX, ctx.attr.local_path)
    download_path = ctx.path(local_path)
    ctx.download(url = ctx.attr.urls, output = local_path)
    ctx.file("WORKSPACE", "workspace(name = \"{name}\")".format(name = ctx.name))
    ctx.file(
        "%s/BUILD" % _DOWNLOAD_PREFIX,
        _BUILD_FILE_TEMPLATE.format(prefix = _DOWNLOAD_PREFIX, path = ctx.attr.local_path)
    )

fetch_file = repository_rule(
    implementation = _fetch_file,
    attrs = {
        "local_path": attr.string(),
        "urls": attr.string_list(mandatory = True),
    },
)
```

### What operating system are you running Bazel on?

Mac OS 10.14.1

### What's the output of `bazel info release`?

`release 0.19.0`

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

N/A

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

N/A (internal repository)

###  Have you found anything relevant by searching the web?

Nope.  Searched stackoverflow and other google searches.  Also the docs.

### Any other information, logs, or outputs that you want to share?

N/A

**Comment by ittaiz** [comment]
Do you have other examples?
Re maven sources example, I’m not 100% sure but, one might argue that you
should know when you generate the dependency, either manually or via a
tool, if sources exist and generate/omit the relevant srcjar url which will
solve the problem

**Comment by davido** [comment]
It always worth checking how Gerrit Code Review project solved a problem you are facing. Because Gerrit is one of the biggest open source projects in the wild that use Bazel. And of course from ca. 180 third party dependencies that Gerrit is using, there are some that do not expose sources on Maven central. 

And yes, we do want to fetch sources per default to provide best experience with IDE integrations for debugging purposes (we generate `.classpath` for Eclipse IDE from Bazel build description and include all sources that we have downloaded using `maven_jar` rule). For that use case we introduced dedicated `maven_jar`'s attribute:

```
  maven_jar(
      name = "foo",
      attach_source = False,
  [...]
```

See https://github.com/GerritCodeReview/gerrit/blob/master/tools/bzl/maven_jar.bzl#L171.

We have extracted our custom `maven_jar` Starlark rule in our own Bazlets repository, to consume them from 100+ Gerrit plugins. So that anyone can easily pull it and use: https://github.com/GerritCodeReview/bazlets/blob/master/tools/maven_jar.bzl .
